src/docfx/lib/schema/JsonSchemaTransformer.cs

```diff
@@ -233,11 +239,28 @@ void TraverseObjectXref(JObject obj, Func<JsonSchema, string, JToken, bool> acti
                     break;
 
                 case JsonSchemaContentType.Xref:
-                    var (xrefError, xrefLink, _, xrefSpec) = context.DependencyResolver.ResolveAbsoluteXref(content, file);
+                    // TODO: the content here must be an UID, not href
+                    var (xrefError, _, _, xrefSpec) = context.DependencyResolver.ResolveAbsoluteXref(content, file);
 
                     if (xrefSpec is InternalXrefSpec internalSpec)
                     {
-                        xrefSpec = internalSpec.ToExternalXrefSpec(context, forXrefMapOutput: false);
+                        if (t_recursionDetector.Value.Contains((content, file)))
+                        {
+                            var referenceMap = t_recursionDetector.Value.Select(x => x.declaringFile).ToList();
+                            referenceMap.Insert(0, file);
+                            throw Errors.CircularReference(referenceMap).ToException();
```

Comment:
move to JsonSchemaTransformer line 93

```suggestion
                            throw Errors.CircularReference(referenceMap, file).ToException();
```